src/cascadia/TerminalControl/SearchBoxControl.cpp

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft Corporation
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "SearchBoxControl.h"
+#include "SearchBoxControl.g.cpp"
+
+using namespace winrt;
+using namespace winrt::Windows::UI::Xaml;
+using namespace winrt::Windows::UI::Core;
+
+namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+{
+    // Constructor
+    SearchBoxControl::SearchBoxControl() :
+        _goForward(false),
+        _isCaseSensitive(false)
+    {
+        InitializeComponent();
+
+        this->CharacterReceived({ this, &SearchBoxControl::_CharacterHandler });
+
+        _goForwardButton = this->FindName(L"SetGoForwardButton").try_as<Controls::Primitives::ToggleButton>();
+        _goBackwardButton = this->FindName(L"SetGoBackwardButton").try_as<Controls::Primitives::ToggleButton>();
+        _textBox = this->FindName(L"TextBox").try_as<Controls::TextBox>();
+
+        auto closeButton = this->FindName(L"CloseButton").try_as<Controls::Button>();
+        auto caseButton = this->FindName(L"CaseSensitivityButton").try_as<Controls::Primitives::ToggleButton>();
+
+        if (closeButton)
+            _focusableElements.insert(closeButton);
+        if (_textBox)
+            _focusableElements.insert(_textBox);
+        if (caseButton)
+            _focusableElements.insert(caseButton);
+        if (_goForwardButton)
+            _focusableElements.insert(_goForwardButton);
+        if (_goBackwardButton)
+            _focusableElements.insert(_goBackwardButton);
+    }
+
+    // Method Description:
+    // - Getter for _goForward
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - bool: the value of _goForward
+    bool SearchBoxControl::GetGoForward()
+    {
+        return _goForward;
+    }
+
+    // Method Description:
+    // - Getter for _isCaseSensitive
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - bool: the value of _isCaseSensitive
+    bool SearchBoxControl::GetIsCaseSensitive()
+    {
+        return _isCaseSensitive;
+    }
+
+    // Method Description:
+    // - Handler for pressing Enter on TextBox, trigger
+    //   text search
+    // Arguments:
+    // - sender: not used
+    // - e: event data
+    // Return Value:
+    // - <none>
+    void SearchBoxControl::QuerySubmitted(winrt::Windows::Foundation::IInspectable const& /*sender*/, Input::KeyRoutedEventArgs const& e)
+    {
+        if (e.OriginalKey() == winrt::Windows::System::VirtualKey::Enter)
+        {
+            auto const state = CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Shift);
+            if (WI_IsFlagSet(state, CoreVirtualKeyStates::Down))
+            {
+                // We do not want the direction flag to change permanately
+                _goForward = !_goForward;
+                _searchEventHandler(*this, _textBox.Text());
+                _goForward = !_goForward;

```

Comment:
discussion: Should the _searchEventHandler simply take an object that has the text and the direction? Changing this "global state" back and forth seems dangerous.

if it is too complicated, you can ignore this comment.

Reply:
It would clean up the logic in this function though.

Reply:
@KaiyuWang16 (I didn't realize you edited my comment to add your response.)

You are correct that it is a big change, but it might let us clean up our code a lot. I leave it up to you.